pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateUserActivityUploadRoute.java

```diff
@@ -0,0 +1,121 @@
+
+        FileUploadService.AuthorizeResult result = TransactionWrapper.withTxn(handle -> {
+            ActivityInstanceDto instanceDto = RouteUtil.findAccessibleInstanceOrHalt(
+                    response, handle, userGuid, studyGuid, instanceGuid, isStudyAdmin);
+
+            ActivityDefStore activityStore = ActivityDefStore.getInstance();
+            ActivityDto activityDto = activityStore.findActivityDto(handle, instanceDto.getActivityId())
+                    .orElseThrow(() -> new DDPException("Could not find activity dto for instance " + instanceGuid));
+            ActivityVersionDto versionDto = activityStore
+                    .findVersionDto(handle, instanceDto.getActivityId(), instanceDto.getCreatedAtMillis())
+                    .orElseThrow(() -> new DDPException("Could not find activity version for instance " + instanceGuid));
+            FormActivityDef activityDef = activityStore.findActivityDef(handle, studyGuid, activityDto, versionDto)
+                    .orElseThrow(() -> new DDPException("Could not find activity definition for instance " + instanceGuid));
+
+            boolean isInstanceReadOnly = ActivityInstanceUtil.isReadonly(
+                    activityDef.getEditTimeoutSec(), instanceDto.getCreatedAtMillis(),
+                    instanceDto.getStatusType().name(), activityDef.isWriteOnce(), instanceDto.getReadonly());
+            if (!isStudyAdmin && isInstanceReadOnly) {
+                String msg = "Activity instance " + instanceGuid + " is read-only, no file upload will be authorized";
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.ACTIVITY_INSTANCE_IS_READONLY, msg));
+            }
+
+            String questionStableId = payload.getQuestionStableId();
+            QuestionDef questionDef = activityDef.getQuestionByStableId(questionStableId);
+            if (questionDef == null) {
+                String msg = "Could not find question with stable id " + questionStableId;
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 404, new ApiError(ErrorCodes.QUESTION_NOT_FOUND, msg));
+            }
+            // todo: check question type is FILE or halt
+
+            boolean isQuestionReadOnly = QuestionUtil.isReadonly(handle, questionDef, instanceDto);
+            if (!isStudyAdmin && isQuestionReadOnly) {
+                String msg = "Question " + questionStableId + " is read-only, no file upload will be authorized";
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.QUESTION_IS_READONLY, msg));
+            }
+
+            User operatorUser = handle.attach(UserDao.class).findUserByGuid(operatorGuid)
+                    .orElseThrow(() -> new DDPException("Could not find operator with guid " + operatorGuid));
+
+            String prefix = String.format("%s/%s/%s", studyGuid, userGuid, instanceDto.getActivityCode());
```

Comment:
I think file uploads are mostly "user-centric", so having a prefix as `study/user/...` seems better for organization and browsing. The rest of prefix is debatable and open to other opinions.